### PR TITLE
FI-3037: SMART User Access Brands Suite

### PR DESCRIPTION
**Pull request by emichaud998**
# Summary
This PR adds a new suite for the SMART v2.2 User Access Brands. This suite contains new tests to validate a publisher's User Access Brands Bundle and it's contained Brand (Organization) and Endpoint resources. Spec tests have also been added for these new tests.

# Testing Guidance
Run spec tests and ensure everything passes, and run locally and make sure everything passes with the preset. Note: Issue with validator validating Brand profile has been filtered out of the validator for now: https://chat.fhir.org/#narrow/stream/291844-FHIR-Validator/topic/SMART.20v2.2E2.20User.20Access.20Brands.3A.20Brand.20validation.20error.3F/near/466321024

### CODE REVIEW

File: .env.production
Comment (Jammjammjamm) [inline .env.production:4]: This needs to be the docker hostname.

File: .env.test
Comment (Jammjammjamm) [inline .env.test:2]: You should make this something fake like the other one to ensure that the tests are never hitting a real service.

File: docker-compose.background.yml
Comment (Jammjammjamm) [inline docker-compose.background.yml:10]: This doesn't belong here.

File: lib/smart_app_launch/smart_access_brands_group.rb
Comment (Jammjammjamm) [inline lib/smart_app_launch/smart_access_brands_group.rb:6]: I'd recommend against adding `_test/group/suite` to the end of ids.

File: lib/smart_app_launch/smart_access_brands_suite.rb
Comment (Jammjammjamm) [inline lib/smart_app_launch/smart_access_brands_suite.rb:45]: The actual filenames should end with `.json.erb`, then you can just delete the suffix.

File: lib/smart_app_launch/smart_access_brands_suite.rb
Comment (arscan) [inline lib/smart_app_launch/smart_access_brands_suite.rb:12]: use the published url (not build.fhir.org) https://hl7.org/fhir/smart-app-launch/brands.html

File: lib/smart_app_launch/smart_access_brands_suite.rb
Comment (arscan) [inline lib/smart_app_launch/smart_access_brands_suite.rb:21]:
```suggestion
      This Test Suite provides conformance testing to ensure the published User-access Brands Bundles are publicly available and
```

File: config/presets/smart_access_brands.json.erb
Comment (arscan) [inline config/presets/smart_access_brands.json.erb:8]: Did we create this example?

I think it might be a good idea to have the examples provided by the guide itself:

* https://hl7.org/fhir/smart-app-launch/Bundle-example1.json
* https://hl7.org/fhir/smart-app-launch/Bundle-example2.json
* https://hl7.org/fhir/smart-app-launch/Bundle-example3.json
* https://hl7.org/fhir/smart-app-launch/Bundle-example4.json 

In the preset, make sure to set the "Endpoint Availability Success Rate (required) *" to "none" for these, as we want to demonstrate that the examples provided are indeed correct (as long as there isn't an expectation that the endpoints are currently accessible).

As far as including the existing one that we made as well, I'm ambivalent.  Does it add anything on top of the examples provided in the IG?  If yes, then I think we should just make sure to title the presets in a way that makes sense (this is an example we came up with... then there are the examples from the IG itself).

Comment (arscan) [reply config/presets/smart_access_brands.json.erb:8]: Hmmmm, I also get this validation error for one of the examples though:  ```Organization.extension[1].extension[1].value.ofType(url): Example URLs are not allowed in this context ```

Hmmm, I think we should just shut that off in the validator.  Checking the format of the URL isn't really relevant to us, as we let the tester choose to actually try to resolve the URLs.

I'll add a comment to the right spot to turn that off in the validator.

File: lib/smart_app_launch/smart_access_brands_suite.rb
Comment (arscan) [inline lib/smart_app_launch/smart_access_brands_suite.rb:32]: Add this in here (per [inferno docs](https://inferno-framework.github.io/inferno-core/docs/Inferno/DSL/FHIRResourceValidation/Validator.html) and [wrapper docs](https://github.com/hapifhir/org.hl7.fhir.validator-wrapper/blob/0093f0b02b9b88c86cf4e1da75d4aed410acf8d5/src/commonMain/resources/static-content/openapi.yml#L218C13-L218C29)):

```ruby
cli_context({
  # Allow example URLs because we give tester option to follow URLs anyhow
  # (configurable) and its nice to be able to have the examples in the IG pass
  allowExampleUrls: true
})
```

Comment (arscan) [reply lib/smart_app_launch/smart_access_brands_suite.rb:32]: That will get rid of this on example2 from the IG
Screenshot 2024-10-01 at 4 17 06 PM

File: lib/smart_app_launch/smart_access_brands_group.rb
Comment (arscan) [inline lib/smart_app_launch/smart_access_brands_group.rb:5]: I think we want this as a `run_as_group`?  I'm not sure it makes much sense to allow users to run the individual groups, as the second relies on the first having been run first, right?

It is nice and defensive, but I'm not sure what the value is.  Plus, the Service Base URL tests are 'run_as_group'.

Screenshot 2024-10-01 at 4 43 12 PM

File: lib/smart_app_launch/smart_access_brands_group.rb
Comment (arscan) [inline lib/smart_app_launch/smart_access_brands_group.rb:10]: Also references build.fhir.org.  Get stable links for all of these.  Because we are specifically pointing to 2.2, use https://hl7.org/fhir/smart-app-launch/STU2.2/ as the base.  I won't comment on all of these... just do a search and fix all of them :)

File: config/presets/smart_access_brands.json.erb
Comment (emichaud998) [reply config/presets/smart_access_brands.json.erb:8]: In terms of whether or not to keep the existing one we made, the only thing it really adds on top of the other examples is that Inferno can host the example endpoints provided in this Bundle and thus do availability validation, whereas these IG examples if we set the success rate input to "None", then there will be no examples that do availability validation. However, I am not sure how much that really matters, so I will leave it up to you whether or not we get rid of it

File: config/presets/smart_access_brands.json.erb
Comment (arscan) [inline config/presets/smart_access_brands.json.erb:7]:
```suggestion
      "name": "smart_access_brands_publication_url",
```

I think you forgot to change this after the update to allow users to paste in bundle contents.

File: lib/smart_app_launch/smart_access_brands_suite.rb
Comment (arscan) [inline lib/smart_app_launch/smart_access_brands_suite.rb:22]: I think we want to also note like this:

>  This test suite is currently designed to fetch and validate a single User-Access Brand Bundle.  It does not currently evaluate the system's ability to allow Health Data Providers to manage all data elements marked “Must-Support” in the “[User Access Brand](http://hl7.org/fhir/smart-app-launch/STU2.2/StructureDefinition-user-access-brand.html)” and “[User Access Endpoint](http://hl7.org/fhir/smart-app-launch/STU2.2/StructureDefinition-user-access-endpoint.html)” profiles.

I bring this up because we aren't doing "must support" checking here, right?  Purely "is this bundle valid".  Which is fine but we should not that, as it is a SHALL we do not test here.

File: config/presets/smart_access_brands.json.erb
Comment (arscan) [reply config/presets/smart_access_brands.json.erb:8]: Good point, let's leave it.

**Review by arscan — APPROVED** [review]
Looks great.  Besides my minor update requests (preset and suite description), some points to just take care of quickly before merge:

The one gap I can think of is that we don't verify that logos resolve. The annoying thing about that is if we do add that capability we will need to add in all that logic to cap the number we verify, and allow the tester to decide to not check if any resolve / at least 1 resolve / blah blah.  So, I think for now we should just note that we do not check these in the suite description, add a ticket, and leave it for a future iteration to look at.  

Are there any other requirements we don't cover?  If so, we should note that in the suite description.  I saw that we don't check MUST SUPPORT but maybe we have others that we chose not to do, in which case just note it and make a ticket that could get picked up in a future iteration.

**Review by arscan — APPROVED** [review]
Changes look excellent, ready to merge!

**Pull request merged**
